### PR TITLE
feat: add bfs and dfs for JS and TS

### PR DESCRIPTION
**Pull request by zhuoqinyue**
If this PR is related to coding or code translation, please read the [contribution guideline](https://github.com/krahets/hello-algo/issues/15#issue-1464069591), fill out the checklist, and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

**Comment by zhuoqinyue** [comment]
hey, K sir. It seems that you foget to check this PR? @krahets

**Comment by krahets** [comment]
The PR related to JS and TS is to be reviewed by @justin-tse.

**Comment by zhuoqinyue** [comment]
OK.

File: codes/javascript/chapter_graph/graph_bfs.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+[v[2], v[5]], [v[3], v[4]], [v[3], v[6]], [v[4], v[5]],
+[v[4], v[7]], [v[5], v[8]], [v[6], v[7]], [v[7], v[8]]];
+const graph = new GraphAdjList(edges);
+console.log("\n初始化后，图为");
+graph.print();
+
+/* 广度优先遍历 BFS */
+const res = graphBfs(graph, v[0]);
+console.log("\n广度优先遍历（BFS）顶点序列为");
+console.log(Vertex.vetsToVals(res));
```
Comment (justin-tse) [inline codes/javascript/chapter_graph/graph_bfs.js:52]:
```suggestion
console.log(Vertex.vetsToVals(res));

```

File: codes/javascript/chapter_graph/graph_dfs.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+const edges = [[v[0], v[1]], [v[0], v[3]], [v[1], v[2]],
+[v[2], v[5]], [v[4], v[5]], [v[5], v[6]]];
+const graph = new GraphAdjList(edges);
+console.log("\n初始化后，图为");
+graph.print();
+
+/* 深度优先遍历 DFS */
+const res = graphDFS(graph, v[0]);
+console.log("\n深度优先遍历（DFS）顶点序列为");
+console.log(Vertex.vetsToVals(res));
```
Comment (justin-tse) [inline codes/javascript/chapter_graph/graph_dfs.js:48]:
```suggestion
console.log(Vertex.vetsToVals(res));

```

File: codes/typescript/chapter_graph/graph_bfs.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+[v[2], v[5]], [v[3], v[4]], [v[3], v[6]], [v[4], v[5]],
+[v[4], v[7]], [v[5], v[8]], [v[6], v[7]], [v[7], v[8]]];
+const graph: GraphAdjList = new GraphAdjList(edges);
+console.log("\n初始化后，图为");
+graph.print();
+
+/* 广度优先遍历 BFS */
+const res: Vertex[] = graphBfs(graph, v[0]);
+console.log("\n广度优先遍历（BFS）顶点序列为");
+console.log(Vertex.vetsToVals(res));
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_bfs.ts:50]:
```suggestion
console.log(Vertex.vetsToVals(res));

```

File: codes/typescript/chapter_graph/graph_dfs.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+const edges: Vertex[][] = [[v[0], v[1]], [v[0], v[3]], [v[1], v[2]],
+[v[2], v[5]], [v[4], v[5]], [v[5], v[6]]];
+const graph: GraphAdjList = new GraphAdjList(edges);
+console.log("\n初始化后，图为");
+graph.print();
+
+/* 深度优先遍历 DFS */
+const res: Vertex[] = graphDFS(graph, v[0]);
+console.log("\n深度优先遍历（DFS）顶点序列为");
+console.log(Vertex.vetsToVals(res));
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_dfs.ts:47]:
```suggestion
console.log(Vertex.vetsToVals(res));

```

File: codes/typescript/chapter_graph/graph_bfs.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+/* 广度优先遍历 BFS */
+// 使用邻接表来表示图，以便获取指定顶点的所有邻接顶点
+const graphBfs = (graph: GraphAdjList, startVet: Vertex): Vertex[] => {
+    // 顶点遍历序列
+    const res: Vertex[] = [];
+    // 哈希表，用于记录已被访问过的顶点
+    const visited: Set<Vertex> = new Set();
+    visited.add(startVet);
+    // 队列用于实现 BFS
+    const que: Vertex[] = [startVet];
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_bfs.ts:20]:
```suggestion
    const que = [startVet];
```

File: codes/typescript/chapter_graph/graph_bfs.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    // 以顶点 vet 为起点，循环直至访问完所有顶点
+    while (que.length) {
+        const vet: Vertex = que.shift();    // 队首顶点出队
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_bfs.ts:23]:
```suggestion
        const vet = que.shift();    // 队首顶点出队
```

File: codes/typescript/chapter_graph/graph_bfs.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            visited.add(adjVet);    // 标记该顶点已被访问
+        }
+    }
+    // 返回顶点遍历序列
+    return res;
+}
+
+/* 初始化无向图 */
+const v: Vertex[] = Vertex.valsToVets([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
+const edges: Vertex[][] = [[v[0], v[1]], [v[0], v[3]], [v[1], v[2]], [v[1], v[4]],
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_bfs.ts:40]:
```suggestion
const v = Vertex.valsToVets([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
const edges = [[v[0], v[1]], [v[0], v[3]], [v[1], v[2]], [v[1], v[4]],
```

File: codes/typescript/chapter_graph/graph_bfs.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+[v[2], v[5]], [v[3], v[4]], [v[3], v[6]], [v[4], v[5]],
+[v[4], v[7]], [v[5], v[8]], [v[6], v[7]], [v[7], v[8]]];
+const graph: GraphAdjList = new GraphAdjList(edges);
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_bfs.ts:43]:
```suggestion
const graph = new GraphAdjList(edges);
```

File: codes/typescript/chapter_graph/graph_bfs.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+console.log("\n初始化后，图为");
+graph.print();
+
+/* 广度优先遍历 BFS */
+const res: Vertex[] = graphBfs(graph, v[0]);
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_bfs.ts:48]:
```suggestion
const res = graphBfs(graph, v[0]);
```

File: codes/typescript/chapter_graph/graph_dfs.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    const res: Vertex[] = [];
+    // 哈希表，用于记录已被访问过的顶点
+    const visited: Set<Vertex> = new Set();
+    dfs(graph, visited, res, startVet);
+    return res;
+}
+
+/* 初始化无向图 */
+const v: Vertex[] = Vertex.valsToVets([0, 1, 2, 3, 4, 5, 6]);
+const edges: Vertex[][] = [[v[0], v[1]], [v[0], v[3]], [v[1], v[2]],
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_dfs.ts:38]:
```suggestion
const v = Vertex.valsToVets([0, 1, 2, 3, 4, 5, 6]);
const edges = [[v[0], v[1]], [v[0], v[3]], [v[1], v[2]],
```

File: codes/typescript/chapter_graph/graph_dfs.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+[v[2], v[5]], [v[4], v[5]], [v[5], v[6]]];
+const graph: GraphAdjList = new GraphAdjList(edges);
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_dfs.ts:40]:
```suggestion
const graph = new GraphAdjList(edges);
```

File: codes/typescript/chapter_graph/graph_dfs.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+console.log("\n初始化后，图为");
+graph.print();
+
+/* 深度优先遍历 DFS */
+const res: Vertex[] = graphDFS(graph, v[0]);
```
Comment (justin-tse) [inline codes/typescript/chapter_graph/graph_dfs.ts:45]:
```suggestion
const res = graphDFS(graph, v[0]);
```

**Review by justin-tse — APPROVED** [review]
@zhuoqinyue Thank you for the clean code.
1. We are used to adding a blank line in the final.
2. In TypeScript, most of the time, less is better, we can use Type inference automatically.

**Comment by zhuoqinyue** [comment]
I've updated it.
But what I'm confused about is why not type constraints on variables in TS.
Are you worried that there will be some `null` or other types here? @justin-tse

**Review by krahets — APPROVED** [review]

**Comment by krahets** [comment]
Hi @zhuoqinyue ! Could you unify the format of the functions in this PR to

```
function hashingSearchLinkedList(map: Map<number, ListNode>, target: number): ListNode | null {
...
```

Now the codes are copied and pasted automatically to the docs based on the recognition by the regular expressions, so we're wondering a consistent naming style of all the functions and classes.

**Comment by zhuoqinyue** [comment]
OK

**Pull request merged**